File: math/ways_to_express_integer_as_sum_of_powers.rb
# @param {Integer} n
# @param {Integer} x
# @return {Integer}
def number_of_ways(n, x)
    mod = 10**9 + 7
    # Find the maximum base such that base**x <= n
    max_base = 1
    while max_base ** x <= n
        max_base += 1
    end
    max_base -= 1

    # dp[i][j]: number of ways to get sum j using bases up to i
    dp = Array.new(max_base + 2) { Array.new(n + 1, 0) }
    # There is 1 way to make sum 0: use nothing
    (0..max_base+1).each { |i| dp[i][0] = 1 }

    (1..max_base).each do |base|
        power = base ** x
        (0..n).each do |sum|
            # Ways without using this base
            dp[base][sum] = dp[base-1][sum]
            # Ways using this base
            if sum >= power
                dp[base][sum] = (dp[base][sum] + dp[base-1][sum-power]) % mod
            end
        end
    end

    dp[max_base][n]
end

File: math/ways_to_express_integer_as_sum_of_powers.py
class Solution:
    def numberOfWays(self, n: int, x: int) -> int:
        mod = 10 ** 9 + 7

        data = {}

        def dp(remaining, current):
            tuple = (remaining, current)
            if tuple in data:
              return data[tuple]

            power = current ** x
            if remaining == 0:
                return 1
            if power > remaining:
                return 0

            take = dp(remaining - power, current + 1)
            skip = dp(remaining, current + 1)

            data[tuple] = (take + skip) % mod
            return data[tuple]

        return dp(n, 1)